draft-ietf-quic-transport.md

```diff
@@ -3981,6 +3979,12 @@ A client MUST NOT include an original connection ID, a stateless reset token, or
 a preferred address.  A server MUST treat receipt of any of these transport
 parameters as a connection error of type TRANSPORT_PARAMETER_ERROR.
 
+max_connection_ids (0x000e):
+
+: The maximum number of connection IDs that the peer is willing to store.
+  This value includes only connection IDs sent in NEW_CONNECTION_ID frames.
```

Comment:
Why "only" connection IDs sent in NCID frames?  The CIDs provided in the handshake and SPA consume space in this list, get retired, etc. If they don't count toward this limit, that would suggest that I'll be temporarily 1-2 over the limit early in the connection, then drift down to the limit as I retire those CIDs. That doesn't seem like the desired effect.

Reply:
Hm, that's true.  But the reverse creates an odd situation where the server can issue 7 CIDs when the client has said they only want 5, but the client can't (or shouldn't) reduce its value by 2 and constrain itself for the lifetime of the connection because of a start-only overflow.

Perhaps the server SHOULD NOT issue more total CIDs than this, but MUST NOT issue more CIDs via NCID?

Reply:
Worst case, from an implementation perspective, if you hit your limit and still have the handshake CIDs, retire them and the right things will happen.

Reply:
Ugh. I don't like including the CID with SPA in this count for the reason @ianswett mentions. Given that the only CID that a client will _store_ is the one that is sent on an SPA, and even that will get used immediately (or get discarded), I think it's reasonable to say that the number of _stored_ CIDs is limited by max_connection_ids.

Reply:
SPA=?

Reply:
I think this discussion is a good reason to not have the enforcement (i.e. close the connection if peer exceeds the maximum).

As @ekr has pointed out in https://github.com/quicwg/base-drafts/issues/1994#issuecomment-441305154, the `max_connection_ids` can be a purely advisory value without defining any error conditions.

Reply:
@dtikhonov Server Preferred Address

Reply:
I'm not sure why enforcing this creates pressure to increase the max. Sure, you need to advertise a value that actually makes sense (i.e. you should request as many CIDs as you think you'll need), and not rely on the peer overflowing this limit (which is exactly what we're trying to avoid in the first place).

Reply:
@kazuho: I think that's fair, and I'm ok with not closing the connection if a peer violates my max_connection_id. We need to be clear that an endpoint that drops a NEW_CID because it has no space MUST NOT send a RETIRE_CID. This is an easy bug, so it's worth noting.

Reply:
If a server has sent a CID as part of SPA, what sequence number does it use for the next NCID?

Reply:
@huitema I think this is unrelated to this PR. We have text about that in https://quicwg.org/base-drafts/draft-ietf-quic-transport.html#issue-cid.

Reply:
It is *not* unrelated. The sequence number is a direct statement on the number of CID out there. I would very much like to have just one number to test. Say, state "sequence number should be lower than the number of retired CID plus the max number of CID accepted by the peer".

Reply:
I think the restriction you're suggesting is a nice simplification Christian.

I'd also like to close the connection if the limit is exceeded, but I know there was some opposition to that in Tokyo that I didn't fully understand.

Reply:
Iff you count the handshake and SPA CIDs, an overflow just from transport parameters is possible if set to 0/1, as @marten-seemann pointed out earlier.  Overflows are also possible based on the advice to replace upon seeing a new CID enter use, though I think that's best addressed by just not recommending/allowing that.

Reply:
@MikeBishop yes indeed, the server always sends at least 1 CID, so a limit of 0 appears to break. But this is easily worked around by only applying the test to NCID frames. This is easier than special-casing the handling of CID[0] and CID[1]/SPA. In particular, the logic for handling RETIRE_CID and creation of new NCID becomes much simpler.

As I mentioned before, a default of 0 does not work very well, because the default for migration disabled is also 0 -- yet enabling migration without any NCID would be a protocol violation. The default should be at least 2 (initial and SPA), and I would prefer 3.